flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateSerializer.java

```diff
@@ -0,0 +1,182 @@
+
+	static ChannelStateByteBuffer wrap(Buffer buffer) {
+		return new ChannelStateByteBuffer() {
+
+			private final ByteBuf byteBuf = buffer.asByteBuf();
+
+			@Override
+			public boolean isWritable() {
+				return byteBuf.isWritable();
+			}
+
+			@Override
+			public int writeBytes(InputStream input, int bytesToRead) throws IOException {
+				return byteBuf.writeBytes(input, Math.min(bytesToRead, byteBuf.writableBytes()));
+			}
+		};
+	}
+
+	static ChannelStateByteBuffer wrap(BufferBuilder bufferBuilder) {
+		final byte[] buf = new byte[1024];
+		return new ChannelStateByteBuffer() {
+			@Override
+			public boolean isWritable() {
+				return !bufferBuilder.isFull();
+			}
+
+			@Override
+			public int writeBytes(InputStream input, int bytesToRead) throws IOException {
+				int left = bytesToRead;
+				for (int toRead = getToRead(left); toRead > 0; toRead = getToRead(left)) {
+					int read = input.read(buf, 0, toRead);
+					int copied = bufferBuilder.append(java.nio.ByteBuffer.wrap(buf, 0, read));
+					Preconditions.checkState(copied == read);
+					left -= read;
+				}
+				bufferBuilder.commit();
+				return bytesToRead - left;
+			}
+
+			private int getToRead(int bytesToRead) {
+				return min(bytesToRead, min(buf.length, bufferBuilder.getWritableBytes()));
+			}
+		};
+	}
+
+	static ChannelStateByteBuffer wrap(byte[] bytes) {
+		return new ChannelStateByteBuffer() {
+			private int written = 0;
+
+			@Override
+			public boolean isWritable() {
+				return written < bytes.length;
+			}
+
+			@Override
+			public int writeBytes(InputStream input, int bytesToRead) throws IOException {
+				final int bytesRead = input.read(bytes, written, bytes.length - written);
+				written += bytesRead;
+				return bytesRead;
+			}
+		};
+	}
+}
+
+class ChannelStateSerializerImpl implements ChannelStateSerializer, ChannelStateDeserializer {
+	private static final int SERIALIZATION_VERSION = 0;
+
+	@Override
+	public void writeHeader(DataOutputStream dataStream) throws IOException {
+		dataStream.writeInt(SERIALIZATION_VERSION);
+	}
+
+	@Override
+	public void writeData(DataOutputStream stream, Buffer... flinkBuffers) throws IOException {
+		stream.writeInt(getSize(flinkBuffers));
+		for (Buffer buffer : flinkBuffers) {
+			ByteBuf nettyByteBuf = buffer.asByteBuf();
+			nettyByteBuf.getBytes(nettyByteBuf.readerIndex(), stream, nettyByteBuf.readableBytes());
+		}
+	}
+
+	private int getSize(Buffer[] buffers) {
+		int len = 0;
+		for (Buffer buffer : buffers) {
+			len = addExact(len, buffer.readableBytes());
+		}
+		return len;
+	}
+
+	@Override
+	public void readHeader(InputStream stream) throws IOException {
+		int version = readInt(stream);
+		Preconditions.checkArgument(version == SERIALIZATION_VERSION, "unsupported version: " + version);
+	}
+
+	/**
+	 * Reads the length of state.
+	 *
+	 * @return size of state in bytes and offset of the current position resulted from reading.
```

Comment:
nit: adjust the description because we only return the size of state in bytes now.